server/src/test/java/org/elasticsearch/repositories/RepositoriesServiceIT.java

```diff
@@ -0,0 +1,102 @@
+
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.Collections;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.not;
+import static org.hamcrest.Matchers.sameInstance;
+
+public class RepositoriesServiceIT extends ESIntegTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return Collections.singletonList(MockRepository.Plugin.class);
+    }
+
+    public void testUpdateRepositoryWithDiffType() {
+        String repositoryName = "test-repo";
+
+        Client client = client();
+        InternalTestCluster cluster = (InternalTestCluster) cluster();
+        RepositoriesService repositoriesService = cluster.getInstance(RepositoriesService.class);
+        Settings settings = cluster.getDefaultSettings();
+        Path location = randomRepoPath();
+
+        Settings.Builder repoSettings = Settings.builder().put(settings).put("location", location);
+
+        PutRepositoryResponse putRepositoryResponse1 =
+            client.admin().cluster().preparePutRepository(repositoryName)
+                .setType(FsRepository.TYPE)
+                .setSettings(repoSettings)
+                .get();
+        assertThat(putRepositoryResponse1.isAcknowledged(), equalTo(true));
+
+        GetRepositoriesResponse getRepositoriesResponse1 =
+            client.admin().cluster().prepareGetRepositories(repositoryName).get();
+
+        assertThat(getRepositoriesResponse1.repositories(), hasSize(1));
+        RepositoryMetaData repositoryMetaData1 = getRepositoriesResponse1.repositories().get(0);
+
+        assertThat(repositoryMetaData1.type(), equalTo(FsRepository.TYPE));
+
+        Repository repository1 = repositoriesService.repository(repositoryName);
+        assertThat(repository1, instanceOf(FsRepository.class));
+
+        // update repository
+
+        boolean diffType = randomBoolean();
+
+        PutRepositoryResponse putRepositoryResponse2 =
+            client.admin().cluster().preparePutRepository(repositoryName)
+                .setType(diffType ? "mock" : FsRepository.TYPE)
+                .setSettings(repoSettings)
+                .get();
+        assertThat(putRepositoryResponse2.isAcknowledged(), equalTo(true));
+
+        GetRepositoriesResponse getRepositoriesResponse2 =
+            client.admin().cluster().prepareGetRepositories(repositoryName).get();
+
+        assertThat(getRepositoriesResponse2.repositories(), hasSize(1));
+        RepositoryMetaData repositoryMetaData2 = getRepositoriesResponse2.repositories().get(0);
+
+        assertThat(repositoryMetaData2.type(), equalTo(diffType ? "mock" : FsRepository.TYPE));
+
+        Repository repository2 = repositoriesService.repository(repositoryName);
+        assertThat(repository2, instanceOf(diffType ? MockRepository.class : FsRepository.class));
```

Comment:
I don't think we need to check the implementation class, instance should be enough.